src/net_processing.cpp

```diff
@@ -1122,34 +1123,40 @@ bool PeerManagerImpl::IsBlockRequested(const uint256& hash)
 
 void PeerManagerImpl::RemoveBlockRequest(const uint256& hash, std::optional<NodeId> from_peer)
 {
-    auto it = mapBlocksInFlight.find(hash);
-    if (it == mapBlocksInFlight.end()) {
-        // Block was not requested
+    auto range = mapBlocksInFlight.equal_range(hash);
+    if (range.first == range.second) {
+        // Block was not requested from any peer
         return;
     }
 
-    auto [node_id, list_it] = it->second;
+    // Currently we don't request more than one peer for same block
+    Assume(mapBlocksInFlight.count(hash) == 1);
 
-    if (from_peer && node_id != *from_peer) {
-        // Block was requested by another peer
-        return;
-    }
+    while (range.first != range.second) {
+        auto [node_id, list_it] = range.first->second;
 
-    CNodeState *state = State(node_id);
-    assert(state != nullptr);
+        if (from_peer && *from_peer != node_id) {
```

Comment:
This is the line valgrind trips over. Adding something like `fprintf(stderr, "from_peer=%ld\n", *from_peer);` above it makes it stop complaining. Maybe this is just a valgrind a bug?